rust-runtime/aws-smithy-http-server/src/request.rs

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+//! Parse HTTP request trait
+
+use aws_smithy_http::body::SdkBody;
+use bytes::Bytes;
+
+/// `ParseHttpRequest` is a generic trait for parsing structured data from HTTP requests.
+///
+/// It is designed to be flexible, because `Input` is unconstrained, it can be used to support
+/// event streams, regular request-response style operations, as well as any other HTTP-based
+/// protocol that we manage to come up with.
+///
+/// It also enables this critical and core trait to avoid being async, and it makes code that uses
```

Comment:
I don't understand these comments in the traits saying that it enables the trait to avoid being `async`.

Reply:
This is a little forward thinking in terms of supporting streaming types. `If` we use the SdkBody type for streaming types, we can avoid async traits as SdkBody type implements directly future polling: https://github.com/awslabs/smithy-rs/blob/main/rust-runtime/aws-smithy-http/src/body.rs#L204

Reply:
that's not totally accurate ^

This trait can avoid being async because _it_ never needs to asynchronously read the body. Instead, it can signal to a driver when the body needs to be written.

However, I'm not sure this is the right design for the server—it doesn't enable any sort of behavior where you'd have a max content length or have parsing interact with whether or not you want to read more of a response body.